src/misc/coloring.c

```diff
@@ -129,6 +133,136 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
+    igraph_color_metadata_t color_metadata; 
+    igraph_integer_t nbr_cnt = igraph_vector_int_size(neighbours);
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++){
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if(!igraph_gen2wheap_has_elem(node_degrees_heap, nbr)){
+            continue;
+        }
+        color_metadata = *((igraph_color_metadata_t*)igraph_gen2wheap_get(node_degrees_heap, nbr));
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            color_metadata.saturation_degree++;
+        }
+        color_metadata.edge_degree--;
+        igraph_gen2wheap_modify(node_degrees_heap, nbr, &color_metadata);
+
+    }
+}
+
+static igraph_integer_t igraph_i_vector_bool_first_true(igraph_vector_bool_t *bool_vector){
+    for(igraph_integer_t index = 0 ; index<igraph_vector_bool_size(bool_vector) ; index++){
+        if(VECTOR(*bool_vector)[index]){
+            return index;
+        }
+    } 
+    return -1;
+}
+static igraph_error_t igraph_i_dsatur_main_loop(
+    const igraph_t *graph, igraph_vector_int_t *colors,
+    igraph_gen2wheap_t *node_degrees_heap, igraph_adjlist_t *adjlist, 
+    igraph_integer_t vc
+) {
+    igraph_integer_t vertices_colored = 0;
+    igraph_vector_bool_t viable_colors;
+    IGRAPH_VECTOR_BOOL_INIT_FINALLY(&viable_colors, vc);
+    while (vertices_colored < vc) {
+        igraph_integer_t node_to_color = igraph_gen2wheap_max_index(node_degrees_heap);
+        igraph_vector_int_t *neighbours = igraph_adjlist_get(adjlist, node_to_color);
+        igraph_integer_t neighbours_count = igraph_vector_int_size(neighbours);
+        igraph_vector_bool_fill(&viable_colors, true);
+        for(igraph_integer_t neighbor_index = 0 ; neighbor_index<neighbours_count ; neighbor_index++){
+            igraph_integer_t neighbour = VECTOR(*neighbours)[neighbor_index]; 
+            if(VECTOR(*colors)[neighbour] != -1){
+                VECTOR(viable_colors)[VECTOR(*colors)[neighbour]] = false;
+            }
+        }
+        igraph_integer_t color = igraph_i_vector_bool_first_true(&viable_colors);
+        igraph_i_dsatur_update_heap(adjlist, node_degrees_heap, neighbours, colors, color);
+        VECTOR(*colors)[node_to_color] = color;
+
+        vertices_colored++;
+    }
+    igraph_vector_bool_destroy(&viable_colors);
+    IGRAPH_FINALLY_CLEAN(1);
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_vertex_coloring_dsatur(
+    const igraph_t *graph, igraph_vector_int_t *colors
+) {
+    igraph_integer_t vc = igraph_vcount(graph);
+
+    IGRAPH_CHECK(igraph_vector_int_resize(colors, vc));
+    igraph_vector_int_fill(colors, -1);   // -1 as a color means uncolored
+
+
+    igraph_adjlist_t adjlist;
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &adjlist, IGRAPH_ALL, IGRAPH_LOOPS_TWICE, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &adjlist);
+    igraph_adjlist_simplify(&adjlist);
```

Comment:
Do we need this, and if yes, why? Where exactly does the algorithm go wrong when using the unsimplified adjlist? Can we handle that without brute-force simplification (which is expensive)?

Reply:
By account for these problems , I mean problems cause by assumption of the graph bring simple broken. A bunch of if statements will have to run in every loop there is.

Reply:
Which will also have a performance hit.

Reply:
I am not asking for changes here. I am asking you why the simplification is necessary. Is this just a pre-emptive measure, just in case the current implementation doesn't work for multigraphs? Or do you know _how exactly_ it fails for multigraphs?

Reply:
It is just a pre-emptive measure when i wrote that comment. Looking at the code thoroughly, only one line will cause problems which is 270
```
       igraph_color_metadata.edge_degree = igraph_vector_int_size(igraph_adjlist_get(&adjlist, vertex));

```
When the heap to store (saturation degree, degree) is being created by inserting each vertex the degree will be calculated incorrectly. So to calculate it correctly, the simple graph  edge degree for each vertex will have to calculated anyways. And if loop and multiple edges are left in adjlist they will only slow the algorithm loop down. So it makes no sense not to do this.

Reply:
by not to do this, I am referring to storing the adjlist as a simple graph